SCMI-Mediator add support for DomUs

oleksiimoisieiev opened the pull request:
This feature is adding SCMI mediator support for DomU. Since toolstack is not using in current rpi5 setup - there is no posibility to test it on rpi5. But this approach was tested on R-Car Gen3 board which is using toolstack to start the Domains.
This pr was made to make SCMI-Mediator support generic.

Config parameters needed to enable scmi feature:
```
CONFIG_ARM_SCI=y                                                                
CONFIG_SCMI_SMC=y                                                               
CONFIG_HOST_DTB_EXPORT=y                                                        
CONFIG_ACCESS_CONTROLLER=y 
```

DomU configuration:
```
arm_sci="scmi_smc"
```

DomU partial dtb:
```
/{
      firmware {
            scmi {
                  scmi_reset:protocol@16 {
                  };
                  scmi_pinctrl:protocol@19 {
                               mux1: mux1 {
                               };
                               mux2: mux2 {
                               };
                  };
      };
      passthrough {
         dev1 {
               resets = <&scmi_reset 1>;
               pinctrl-0 = <&mux1>;
         };
         dev2 {
               pinctrl-0 = <&mux2>;
         };
      };
};
 ```
TODO:
- after review make a PR to xen-4.19-xt
Postponed:
- add new syscall to flask.

oleksiimoisieiev committed:
tools/arm: add "arm_sci" option to xl.cfg

This enumeration sets SCI type for the domain. Currently there is
two possible options: either 'none' or 'scmi_smc'.

'none' is the default value and it disables SCI support at all.

'scmi_smc' enables access to the Firmware from the domains using SCMI
protocol and SMC as transport.

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

oleksiimoisieiev committed:
xen/arm: add SCI mediator support for DomUs

Integration of the SCMI mediator with xen libs:
- add hypercalls to aquire SCI channel and set device permissions
for DomUs;
- add SCMI_SMC nodes to DomUs device-tree based on partial device-tree;
- SCI requests redirection from DomUs to Firmware.

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

oleksiimoisieiev committed:
libs: libxenhypfs - handle blob properties

libxenhypfs will return blob properties as is. This output can be used
to retrieve information from the hypfs. Caller is responsible for
parsing property value.

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

oleksiimoisieiev committed:
xen/arm: Export host device-tree to hypfs

If enabled, host device-tree will be exported to hypfs and can be
accessed through /devicetree path.
Exported device-tree has the same format, as the device-tree
exported to the sysfs by the Linux kernel.
This is useful when XEN toolstack needs an access to the host device-tree.

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

oleksiimoisieiev committed:
xen: arch: scmi_smc: remove memory mapping for hardware domain

Mapping for the hardware domain should be done during domain
construciton as it is done for other domains.

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

oleksiimoisieiev committed:
xen: domain_build: Integration of the SCMI mediator to Domain-0

Integration of the SCMI-Mediator feature to the Domain-0 construction
process. It includes shared memory node creation and mapping with
phandle update on the scmi node.
When creating hardware domain there is a need to set correct phandle
to the "shmem" property on the scmi node. Shared memory node is
generated while scmi node is copied as is from Xen device-tree.
Correct shmem phandle is set during domain device-tree processing.

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc commented on xen/arch/arm/domain.c:
```diff
@@ -698,6 +702,27 @@ int arch_sanitise_domain_config(struct xen_domctl_createdomain *config)
 }
 
+#ifdef CONFIG_HOST_DTB_EXPORT
+#define HYPFS_PROPERTY_MAX_SIZE 4096
```
I believe that DTB is much bigger than 4k. Also, naming is quite ambiguous. It just says "hypfs property". Which property exactly? 

I believe, this is the leftover from your previous version. But in this version it makes no sense

oleksiimoisieiev replied:
Sorry. it's leftover. I've made HOST_DTB_MAX_SIZE config parameter

oleksiimoisieiev replied:
with default as 8192. but probably 4096 is enough because current dtb in rcar is 2150

lorc commented on xen/arch/arm/domain.c:
```diff
@@ -698,6 +702,27 @@ int arch_sanitise_domain_config(struct xen_domctl_createdomain *config)
     return 0;
 }
 
+#ifdef CONFIG_HOST_DTB_EXPORT
```
Why it is living in the middle of `domain.c`?  This file is for low-level domain state manipulation

oleksiimoisieiev replied:
Couldn't find a better place. What do you suggest?

lorc replied:
`bootfdt.c` maybe?

lorc commented on xen/arch/arm/domain_build.c:
```diff
@@ -1741,6 +1755,12 @@ static int __init handle_node(struct domain *d, struct kernel_info *kinfo,
 
+#ifdef CONFIG_SCMI_SMC
+        res = mem_permit_access(kinfo->d, kinfo->d->arch.sci_channel.paddr,
+                0x1000);
```
Please no magic values

oleksiimoisieiev replied:
fixed

lorc replied:
There is still constant 0x1000

lorc commented on xen/arch/arm/domain_build.c:
```diff
@@ -1590,6 +1590,20 @@ int __init make_chosen_node(const struct kernel_info *kinfo)
+    if ( res )
+        return res;
+
+    return map_regions_p2mt(d, gaddr_to_gfn(addr), PFN_DOWN(len),
```
Should you remove iomem access permission if this call is failed?

oleksiimoisieiev replied:
added

lorc replied:
I don't see where it was fixed.

oleksiimoisieiev replied:
Merged these changes to wrong commit. I'm sorry... now fixec

lorc commented on xen/include/xen/scmi_dt_maker.h:
```diff
@@ -16,10 +16,13 @@ int __init scmi_dt_make_shmem_node(struct kernel_info *kinfo);
 int __init scmi_dt_scan_node(struct kernel_info *kinfo, void *pfdt,
                              int nodeoff);
+int __init scmi_dt_set_phandle(struct kernel_info *kinfo,
+        const char *name);
```
formatting is not consistent with the previous lines.

oleksiimoisieiev replied:
fixed

lorc commented on xen/drivers/passthrough/arm/scmi_dt_maker.c:
```diff
@@ -311,3 +311,16 @@ int __init scmi_dt_scan_node(struct kernel_info *kinfo, void *pfdt,
     return rc;
 }
+
+int __init scmi_dt_set_phandle(struct kernel_info *kinfo,
```
It is strange to add a function, but not to call it. Could this commit be squashed into one that really uses this function?

oleksiimoisieiev replied:
squashed

lorc commented on xen/arch/arm/domain_build.c:
```diff
@@ -1799,6 +1811,15 @@ static int __init handle_node(struct domain *d, struct kernel_info *kinfo,
+#ifdef CONFIG_SCMI_SMC
+    if ( dt_match_node(scmi_matches, node) )
+    {
+        res = scmi_dt_set_phandle(kinfo, dt_node_full_name(node));
```
Why are you passing node name  to that function and then resolving node back? Why you can't pass `node` as is?

oleksiimoisieiev replied:
because it's taken from different device tree

lorc commented on xen/drivers/passthrough/device_tree.c:
```diff
@@ -318,6 +319,10 @@ int iommu_do_dt_domctl(struct xen_domctl *domctl, struct domain *d,
             break;
         }
 
+        ret = ac_assign_dt_device(dev, d);
```
Commit message does not mention this change. Are you sure it should be in this patch?

oleksiimoisieiev replied:
moved to separate patch

lorc commented on tools/libs/light/libxl_arm.c:
```diff
@@ -1568,12 +1575,277 @@ static int copy_node_by_path(libxl__gc *gc, const char *path,
+    uint64_t _guest_pfn = guest_addr >> XC_PAGE_SHIFT;
+
+    assert(paddr && guest_addr);
+    LOG(DEBUG, "iomem %"PRIx64, _paddr_pfn);
```
debug message could be clearer, as this is not iomem per se

oleksiimoisieiev replied:
reworded

lorc commented on tools/libs/light/libxl_arm.c:
```diff
@@ -1568,12 +1575,277 @@ static int copy_node_by_path(libxl__gc *gc, const char *path,
+    uint32_t phandle;
+
+    node_next = fdt_first_subnode(pfdt, nodeoff);
+    while ( node_next > 0 )
```
code style

oleksiimoisieiev replied:
+

lorc commented on tools/libs/light/libxl_arm.c:
```diff
@@ -1568,12 +1575,277 @@ static int copy_node_by_path(libxl__gc *gc, const char *path,
+
+        phandle = fdt_get_phandle(pfdt, node_next);
+
+        rc = fdt_get_path(pfdt, node_next, full_name, 128);
```
Are you sure that 128 bytes will be enough for node path? AFAIK DTS specification does not guarantee this

oleksiimoisieiev replied:
from my standpoint it's enough because it is expected to work only with scmi nodes and it's child nodes. As for me I see no point to allocate it dynamically.

lorc replied:
At least use `sizeof(full_name)` here. No need for magic numbers.

lorc commented on tools/libs/light/libxl_arm.c:
```diff
@@ -1568,12 +1575,277 @@ static int copy_node_by_path(libxl__gc *gc, const char *path,
+        rc = fdt_begin_node(fdt, name_from_path(full_name));
+        if (rc) return rc;
+
+        rc = dt_copy_properties(gc, fdt, xen_fdt, full_name);
```
Why are you passing `full_name` to this function? It uses ` get_path_nodeoff` to basically get `node_next` value back. Why do you need this extra step?

oleksiimoisieiev replied:
this is essential because there node_offset of the caller is related to pfdt while in dt_copy_properties I need to get nodeoffset of xen_dtb. There are 3 fdt entities I have:
pfdt - partial device tree
fdt - domain device tree
xen_fdt - xen device_tree from hypfs

lorc commented on tools/libs/light/libxl_arm.c:
```diff
@@ -1568,12 +1575,277 @@ static int copy_node_by_path(libxl__gc *gc, const char *path,
+    int r;
+
+    if (fdt_magic(fdt) != FDT_MAGIC) {
+         LOG(ERROR, "Partial FDT is not a valid Flat Device Tree");
```
This is not the partial device tree

oleksiimoisieiev replied:
fixed

lorc commented on tools/libs/light/libxl_arm.c:
```diff
@@ -1568,12 +1575,277 @@ static int copy_node_by_path(libxl__gc *gc, const char *path,
+
+     r = fdt_check_header(fdt);
+     if (r) {
+         LOG(ERROR, "Failed to check the partial FDT (%d)", r);
```
This is not the partial device tree

oleksiimoisieiev replied:
fixed

lorc commented on tools/libs/light/libxl_arm.c:
```diff
@@ -1568,12 +1575,277 @@ static int copy_node_by_path(libxl__gc *gc, const char *path,
+        goto out;
+    }
+
+    nodeoff = get_path_nodeoff("/scmi", pfdt);
```
Should it be `/firmware/scmi` ? I believe we discussed this at the daily call

oleksiimoisieiev replied:
yup. sorry. it's leftover

lorc commented on tools/libs/light/libxl_arm.c:
```diff
@@ -1568,12 +1575,277 @@ static int copy_node_by_path(libxl__gc *gc, const char *path,
+    }
+
+    nodeoff = get_path_nodeoff("/scmi", pfdt);
+    if ( nodeoff <= 0 ) {
```
coding style

oleksiimoisieiev replied:
2fixed

lorc commented on xen/arch/arm/domctl.c:
```diff
@@ -48,6 +49,16 @@ static int handle_vuart_init(struct domain *d,
+    sci_info->paddr = d->arch.sci_channel.paddr;
+    sci_info->func_id = d->arch.sci_channel.guest_func_id;
+#endif /* CONFIG_ARM_SCI */
+    return 0;
```
I believe you need to return `-ENODEV` if `CONFIG_ARM_SCI` is not enabled or SCI is not enabled for the domain

oleksiimoisieiev replied:
fixed

lorc replied:
What about the second part of my comment? What if SCI is not enabled for that particular domain? It will have no valid paddr and func id in that case

oleksiimoisieiev commented on xen/arch/arm/domctl.c:
```diff
@@ -48,6 +49,16 @@ static int handle_vuart_init(struct domain *d,
+    return 0;
+}
+
+
```
remove extra enter

oleksiimoisieiev replied:
fixed

oleksiimoisieiev commented on xen/arch/arm/domctl.c:
```diff
@@ -176,9 +187,20 @@ long arch_do_domctl(struct xen_domctl *domctl, struct domain *d,
+    {
         return subarch_do_domctl(domctl, d, u_domctl);
     }
+    }
```
remove leftover

oleksiimoisieiev replied:
fixed

oleksiimoisieiev committed:
xen: scmi_dt_maker: Fix code style minor issue

Remove space before function definition.

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

oleksiimoisieiev committed:
xen: device_tree: Assign devices to the Domain-0 during build

Assgin devices to the access-controller during Domain-0 build.
This registers devices in the SCMI server for the Domain-0.

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

oleksiimoisieiev commented:
@lorc @GrygiriiS updated. ready to stage 2

lorc commented:
I see no new changes

oleksiimoisieiev commented:
I'm really sorry... pushed to another branch... Updated @lorc

lorc reviewed (commented):
Looks like you are missing some fixes, although you said that you fixed it

oleksiimoisieiev commented:
Fixed all. Merged changes to the wrong commit. I think once Review is finished. I will make a pr to 4.19-xt0.1 with already reviewed code

lorc reviewed (commented):
With the last two comments resolved:

`Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>`

oleksiimoisieiev commented:
In reply to comment: https://github.com/xen-troops/xen/pull/214#discussion_r1786146765
I'm returning -ENODEV in this case.

lorc commented:
> I'm returning -ENODEV in this case.

Where?

Imagine you have purely virtual DomU without access to hardware. It does not need access to SCMI and SCMI should be disabled for this domain. In this case this domctl should return EINVAL or ENODEV or some other error.

oleksiimoisieiev commented on xen/arch/arm/domctl.c:
```diff
@@ -48,6 +49,17 @@ static int handle_vuart_init(struct domain *d,
+    sci_info->func_id = d->arch.sci_channel.guest_func_id;
+    return 0;
+#else
+    return -ENODEV;
```
This is some weird github twist... probably.

oleksiimoisieiev commented:
https://github.com/xen-troops/xen/pull/214#discussion_r1786159475

fixed.
Adding rb
